File: java/json-transform/src/main/java/co/nlighten/jsontransform/functions/TransformerFunctionUrlEncode.java
package co.nlighten.jsontransform.functions;

import co.nlighten.jsontransform.functions.common.*;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.UnsupportedEncodingException;
import java.net.URLEncoder;
import java.util.Map;

/*
 * For tests
 * @see TransformerFunctionUrlEncodeTest
 */
public class TransformerFunctionUrlEncode extends TransformerFunction {
    static final Logger log = LoggerFactory.getLogger(TransformerFunctionUrlEncode.class);

    public TransformerFunctionUrlEncode() {
        super(FunctionDescription.of(
                Map.of(
                        "charset", ArgumentType.of(ArgType.Enum).position(0).defaultEnum("UTF-8")
                )
        ));
    }
    @Override
    public Object apply(FunctionContext context) {
        var str = context.getString(null);
        if (str == null) {
            return null;
        }
        try {
            var charset = context.getEnum("charset");
            return URLEncoder.encode(str, charset);
        } catch (UnsupportedEncodingException e) {
            log.warn(context.getAlias() + " failed", e);
            return null;
        }
    }
}

File: javascript/json-transform/src/functions/TransformerFunctionUrlEncode.ts
import TransformerFunction from "./common/TransformerFunction";
import FunctionContext from "./common/FunctionContext";
import { ArgType } from "./common/ArgType";

class TransformerFunctionUrlEncode extends TransformerFunction {
  constructor() {
    super({
      arguments: {
        charset: { type: ArgType.Enum, position: 0, defaultEnum: "UTF-8" },
      },
    });
  }

  override async apply(context: FunctionContext): Promise<any> {
    const str = await context.getString(null);
    if (str == null) {
      return null;
    }
    try {
      const charset = await context.getEnum("charset");
      if (charset?.startsWith("UTF-16")) {
        throw new Error("TODO: not implemented"); // TODO: not implemented
      }
      return encodeURIComponent(str).replace(/%20/g, "+");
    } catch (e: any) {
      console.warn(context.getAlias() + " failed", e);
      return null;
    }
  }
}

export default TransformerFunctionUrlEncode;
